### PR TITLE
#228 Update timestamps when trimming duplicate responses

### PR DESCRIPTION
**Pull request by CarlosNZ**
Implements #228.

Changed the field name on application_response and review_response from `time_created` to `time_updated`

Also updated tests. Can run just these tests (as opposed to whole test suite) by running `yarn test` from within the plugin folder.

Front-end PR: https://github.com/openmsupply/application-manager-web-app/pull/422
~~Will need some front-end adjustments for the new field name, will add that in front-end PR shortly.~~

### CODE REVIEW

File: src/plugins/action_trim_responses/src/databaseMethods.ts [hunk 1]
```diff
@@ -64,6 +64,38 @@ const databaseMethods = (DBConnect: any) => ({
       throw err
     }
   },
+  updateApplicationResponseTimestamps: async (responsesToUpdate: number[]) => {
+    const text = `UPDATE application_response
+      SET time_updated = current_timestamp
```
Comment (andreievg) [inline src/plugins/action_trim_responses/src/databaseMethods.ts:69]: I think we discussed for this to be matched to latest application/review status history ?

**Review by andreievg — CHANGES_REQUESTED** [review]
Ohh man, i've thought of some use cases where this would need to change (and would probably need different actions for application_response trimming and review_response trimming)

I think the logic should be

**Application Response** (only changes I think is the deletion of empty response)
* get application_responses for application
* order by timestamp
* group in for each (like you've done) by template_element_id
* compare two of the latest (if match remove the latest)
* if latest is empty response remove it <- i think this is new

**Review Response**
* get all review_response for review
* order by timestamp
* group by either application_response_id or review_response_id (whichever exists for review_response) <- this is new
* compare two of the latest (if match remove the latest)
* if latest is empty response remove  it <- i think this is new

File: src/plugins/action_trim_responses/src/databaseMethods.ts (hunk 1 above)
Comment (CarlosNZ) [reply src/plugins/action_trim_responses/src/databaseMethods.ts:69]: Oh, I don't remember discussing that, sorry. That assumes the update status Action (to create a new status_history) would be ahead of the `trimResponses` Action. Is that a reasonable assumption? 
How about the changeStatus Action pass along the new status timestamp in its output, then we can make this Action use that timestamp if it exists, but write the current_timestamp if it doesn't?

Comment (andreievg) [reply src/plugins/action_trim_responses/src/databaseMethods.ts:69]: sure, or query for it in the update statement. up to you.

> That assumes the update status Action (to create a new status_history) would be ahead of the trimResponses

in the set `is visible` PR I'll make one empty template with 'core actions' (we are getting a few of them now, want to keep track of them)

Comment (CarlosNZ) [reply src/plugins/action_trim_responses/src/databaseMethods.ts:69]: Yeah, I've been trying to keep a list of the core/default actions in #184

Comment (CarlosNZ) [reply src/plugins/action_trim_responses/src/databaseMethods.ts]: Okay, changes to reflect the above: 65197f97

Haven't done extensive testing yet, but it passes the existing tests (which only deals with Level 1 reviews)

File: src/plugins/action_trim_responses/src/databaseMethods.ts [hunk 2]
```diff
@@ -64,6 +66,38 @@ const databaseMethods = (DBConnect: any) => ({
...
+    } catch (err) {
+      console.log(err.message)
+      throw err
+    }
+  },
+  updateReviewResponseTimestamps: async (responsesToUpdate: number[], timestamp: string) => {
+    const text = `UPDATE review_response
+      SET time_updated = $1
+      WHERE id = ANY ($2)
+      RETURNING (SELECT code FROM
```
Comment (andreievg) [inline src/plugins/action_trim_responses/src/databaseMethods.ts:88]: hmm why do we need 'code' from application_response ? (maybe from template_element ?)

File: src/plugins/action_trim_responses/src/trimResponses.ts [hunk 3]
```diff
@@ -1,63 +1,77 @@
...
+      ? 'review_response_link_id'
+      : 'application_response_id'
+
+    const responsesById = groupResponses(responses, groupByField)
     const responsesToDelete: number[] = []
+    const responsesToUpdate: number[] = []
 
-    Object.values(responsesByCode).forEach((responseArray) => {
+    Object.values(responsesById).forEach((responseArray) => {
       if (responseArray.length === 1) return
```
Comment (andreievg) [inline src/plugins/action_trim_responses/src/trimResponses.ts:47]: can you add an 'undefined' review_response to the test and see if it's deleted please ?

use case -> reviewer starts review, marks one question as declined, all others should be deleted

File: src/plugins/action_trim_responses/src/trimResponses.ts [hunk 4]
```diff
@@ -1,63 +1,77 @@
...
-      if (!(code in responsesByCode)) responsesByCode[code] = [response]
-      else responsesByCode[code].push(response)
-    })
+    const reviewLevel = responses?.[0]?.level
+
+    // Create object of indexed responses, with array of response
+    // objects for each indexed id
 
+    const groupByField = !reviewId
+      ? 'template_element_id'
```
Comment (andreievg) [inline src/plugins/action_trim_responses/src/trimResponses.ts:37]: this is stylistic (so not required change, I've suggested this to Nicole a few times), but more nested in-line condition statement are much harder to read then just if statement.

File: database/insertData/03_template_D_basicReviewTesting.js [hunk 5]
```diff
@@ -359,14 +359,30 @@ exports.queries = [
...
+                trigger: ON_REVIEW_SUBMIT
+                sequence: 2
+                parameterQueries: {
+                  reviewId: {
+                    operator: "objectProperties"
+                    children: ["applicationData.record_id"]
+                  }
+                  timestamp: {
+                    operator: "objectProperties"
+                    children: ["output.reviewStatusHistoryTimestamp"]
```
Comment (andreievg) [inline database/insertData/03_template_D_basicReviewTesting.js:384]: ideally we would also do this for ON_APPLICATION_SUBMIT

**Review by andreievg — CHANGES_REQUESTED** [review]
Look good, thanks Carl !

One remaining question (as per below). Would it delete undefined review_response ?

File: src/plugins/action_trim_responses/src/databaseMethods.ts (hunk 2 above)
Comment (CarlosNZ) [reply src/plugins/action_trim_responses/src/databaseMethods.ts:88]: It's `code` from application_response JOIN template_element -- so it is the template_element code ;)

File: src/plugins/action_trim_responses/src/trimResponses.ts (hunk 4 above)
Comment (CarlosNZ) [reply src/plugins/action_trim_responses/src/trimResponses.ts:37]: Yeah, I agree in general, although when the expressions are short I quite like the way Prettier formats them to be quite logical. But yeah, definitely when the expressions are long -- like chunks of JSX! ;)

File: database/insertData/03_template_D_basicReviewTesting.js (hunk 5 above)
Comment (CarlosNZ) [reply database/insertData/03_template_D_basicReviewTesting.js:399]: ef4d5dc0

File: src/plugins/action_trim_responses/src/trimResponses.ts (hunk 3 above)
Comment (CarlosNZ) [reply src/plugins/action_trim_responses/src/trimResponses.ts:47]: Okay, that should be deleting NULL responses now: ef4d5dc0

Comment (andreievg) [reply src/plugins/action_trim_responses/src/trimResponses.ts:47]: I think this will still not quite work if there is one review_response

**Review by andreievg — CHANGES_REQUESTED** [review]

File: src/plugins/action_trim_responses/src/trimResponses.ts (hunk 3 above)
Comment (CarlosNZ) [reply src/plugins/action_trim_responses/src/trimResponses.ts]: Yeah good point. Think it should be correct now: d0aad32f

Just to clarify -- if a response has no duplicate (i.e. it's perId array will have one element), that element's timestamp should be updated too.

**Review by andreievg — APPROVED** [review]
thanks, look good!

**Pull request merged**